src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/SqlCommandReliabilityTest.cs

```diff
@@ -524,160 +526,164 @@ public async void RetryExecuteAsyncCancel(string cnnString, SqlRetryLogicBasePro
+            await ProcessDataAsAsync(cnnString, provider, query, concurrentExecution, cmd => cmd.ExecuteReaderAsync());
+            await ProcessDataAsAsync(cnnString, provider, query, concurrentExecution, cmd => cmd.ExecuteXmlReaderAsync());
+
+            if (DataTestUtility.IsNotAzureSynapse())
             {
-                using (SqlConnection cnn = new SqlConnection(cnnString))
-                using (SqlCommand cmd = cnn.CreateCommand())
-                {
-                    cnn.Open();
-                    cmd.RetryLogicProvider = provider;
-                    cmd.CommandText = query;
-                    Assert.Throws<AggregateException>(() => cmd.ExecuteNonQuery());
-                }
-            });
-            Assert.Equal(numberOfTries * concurrentExecution, retriesCount + concurrentExecution);
+                query += " FOR XML AUTO";
+                await ProcessDataAsAsync(cnnString, provider, query, concurrentExecution, cmd => cmd.ExecuteXmlReaderAsync());
+            }
+        }
+        #endregion
 
-            retriesCount = 0;
-            Parallel.For(0, concurrentExecution,
-            i =>
+        #region private members
+        private SqlCommand CreateCommand(SqlConnection cnn, SqlRetryLogicBaseProvider provider, int cancelAfterRetries)
+        {
+            cnn.Open();
+            SqlCommand cmd = cnn.CreateCommand();
+            cmd.RetryLogicProvider = provider;
+            cmd.RetryLogicProvider.Retrying += (object s, SqlRetryingEventArgs e) =>
             {
-                using (SqlConnection cnn = new SqlConnection(cnnString))
-                using (SqlCommand cmd = cnn.CreateCommand())
+                Assert.Equal(e.RetryCount, e.Exceptions.Count);
+                Assert.NotEqual(TimeSpan.Zero, e.Delay);
+
+                if (e.RetryCount >= cancelAfterRetries)
                 {
-                    cnn.Open();
-                    cmd.RetryLogicProvider = provider;
-                    cmd.CommandText = query;
-                    Assert.Throws<AggregateException>(() => cmd.ExecuteReader());
+                    e.Cancel = true;
                 }
-            });
-            Assert.Equal(numberOfTries * concurrentExecution, retriesCount + concurrentExecution);
+            };
+            return cmd;
+        }
 
-            if(DataTestUtility.IsNotAzureSynapse())
+        private static void ProcessDataInParallel(string cnnString, SqlRetryLogicBaseProvider provider,
+                                                  string query, int concurrentExecution,
+                                                  Action<SqlCommand> commandAction)
+        {
+            int numberOfTries = provider.RetryLogic.NumberOfTries;
+            int delayCount = 10_000;
+            int retriesCount = 0;
+            var exceptions = new ConcurrentQueue<Exception>();
+
+            ThreadPool.SetMinThreads(concurrentExecution * 2, concurrentExecution * 2);
```

Comment:
I don't think this is the best place to SetMinThreads. SetMinThreads affects the entire process and we need to return it to its previous value to avoid affecting subsequent tests. I'd put SetMinThreads in a constructor for the test class, make the class IDisposable, and set min threads back in the Dispose() method.

Same for line 641.